### DISCUSSION TITLE
`app()->terminating()` not (always) called with `queue:work`

### DISCUSSION THREAD

**Discussion by fideloper** [laravel/framework#32779] [General]
I'm investigating doing some cleanup on some temp files while not relying on `__destruct()` (which sometimes is called too early).

I found the following. It's not necessarily an issue but was interesting/a bit inconsistent.

I have a queue job that registers a callback with `app()->terminating()`:

```php
app()->terminating(function() {
    Log::info("{$this->job->getJobId()} is running on app termination");
});
```

**Here's what I found:**

* ✅ `queue:listen` gets a log line for every job, which makes sense as `queue:listen` restarts the framework for each job
*  ✅ `queue:work --once` gets a log line for every job since it fires once and then allows the framework to stop (no `while (true)` loop)
* ❌`queue:work` - Does **NOT** get a log line, ever. This is because `Illuminate\Queue\Worker` uses [`exit($status)` to stop the worker in all cases](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Queue/Worker.php#L609-L637) (graceful shutdown and ungraceful). This stops the framework from doing regular shutdown behavior.

This is obviously not an issue for most, but since I ran into it, I wanted to see if it was worth changing. 

The most obvious solution would be to call `app()->terminate()` just before exiting. This seems to be exactly what happens for web requests, so it doesn't feel like it would miss too much in terms of shutdown behavior.

Note that the [Console Kernel object also calls `app()->terminate()`](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Foundation/Console/Kernel.php#L148) so all other artisan calls are terminating the app. 

**It's really just `queue:work` that's circumventing normal behavior.** Perhaps the queue worker could stop by `break`'ing out of the `while (true)` loop instead of exiting.

> There could be issues of exiting with the correct status code there, so another idea is to call app()->terminate() just before calling `exit($status);`

**Comment by taylorotwell** [comment]
Seems like a reasonable change.

**Comment by fideloper** [comment]
I’ll try for a PR this afternoon (ish) 👍🏻

**Reply by david-kristin** [reply]
Hi, @fideloper, any progress on this one?

**Reply by fideloper** [reply]
You can see the (not-merged) PR here: https://github.com/laravel/framework/pull/32889 and some related PR's that were later merged in that thread.

**Reply by fideloper** [reply]
(Specifically, this PR was later merged: https://github.com/laravel/framework/pull/32889 )
